packages/realm-web-integration-tests/src/remote-mongodb-service.test.ts

```diff
@@ -319,4 +322,101 @@ describe("Remote MongoDB", () => {
         expect(result.modifiedCount).equals(0);
         expect(typeof result.upsertedId).equals("object");
     });
+
+    it("can watch changes correctly", async function () {
+        this.timeout(10_000);
+        this.slow(2_000);
+
+        const mongodb = app.services.mongodb("local-mongodb");
+        const db = mongodb.db("test-database");
+        const collection = db.collection("test-collection");
+
+        await collection.deleteMany({});
+
+        const sleep = async (time: number) =>
+            new Promise(resolve => setInterval(resolve, time));
+        const str =
+            "use some odd chars to force weird encoding %\n\r\n\\????>>>>";
+        await Promise.all([
+            (async () => {
+                // There is a race with creating the watch() streams, since they won't
+                // see inserts from before they are created.
+                // Wait 500ms (490+10) before first insert to try to avoid it.
+                await sleep(490);
+                for (let i = 0; i < 10; i++) {
+                    await sleep(10);
+                    await collection.insertOne({ _id: i, hello: "world", str });
+                }
+                await collection.insertOne({ _id: "done", done: true }); // break other sides out of loop
+            })(),
+            // Watch any event
+            (async () => {
+                let expected = 0;
+                for await (const event of collection.watch()) {
+                    if (event.operationType === "insert") {
+                        if (event.fullDocument.done) break;
+                        expect(event.fullDocument._id).equals(expected++);
+                    } else {
+                        throw new Error(
+                            `Unexpected ${event.operationType} event`,
+                        );
+                    }
+                }
+                expect(expected).equals(10);
+            })(),
+            // Watch using the filter option
+            (async () => {
+                const filter = {
+                    $or: [
+                        { "fullDocument._id": 3, "fullDocument.str": str },
+                        { "fullDocument.done": true },
+                    ],
+                };
+                let seenIt = false;
+                for await (const event of collection.watch({ filter })) {
```

Comment:
This is using the async generator API which might be too exotic or not a good fit with a React app - see https://jira.mongodb.org/browse/RJS-462

Reply:
+1

Reply:
This is tedious to use in non async functions

Reply:
but I like how @RedBeard0531 remembers to use these features, cause for me this does not even exist :)

Reply:
Personally I feel we should release as is, to get the feature out (users have been waiting for this quite a while) and then actively pursue feedback on this API.

I don't think it would be hard to change this into an event emitter, but it would have to be done for Realm JS as well, which would take a bit of time too.

Reply:
> Does it add anything over an event emitter? (I see/share your scepticism).

If users are just iterating the events as they come in from the server, I do think it's a simpler API to consume (once you know and understand the syntax).

> How does one unsubscribe / unwatch?

Thats a good question - I don't see any code explicitly closing the request to the server.
Reading both https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function* it's unclear to me how the caller of a generator would signal to the implementer of the generator that they're done using it.

If it's not possible to stop consuming a watch, that's for sure a significant downside to this API.

Reply:
Here's a blog post that proposes an approach for cancelling iteration precisely in relation to SSEs: http://seg.phault.net/blog/2018/03/async-iterators-cancellation/

Reply:
What's the alternative API and how long would it take to change it?
Are you all in agreement the API is too exotic or should we list up the pro/con for alternatives?

Reply:
These are some of our alternative, off the top of my head:

# Alternatives

## 0. Keep the `AsyncIterator` as is.

Whatever we do, I believe it might make sense to keep this API around for expert users. I do see the advantages of the pull-based iteration which this API enables and I think it could be valuable for an advanced user of the library.

## 1. A simple listener pattern, like the Node.js driver `ChangeStream` API

One alternative is the simplified listener pattern which the Node.js driver is currently exposing: https://mongodb.github.io/node-mongodb-native/3.6/api/ChangeStream.html

This is probably my favorite, since it's simpler to implement and it mimics the Node.js driver enabling transfer of knowledge between drivers and SDKs.

```javascript
const stream = collection.watch();
stream.onNext(change => {
  // Do whatever with the change event ...
});
// After some time, when we're done with the stream
stream.close()
```

## 2. A full-on listener pattern through an API similar to (but perhaps a bit more limited than) [EventEmitter](https://nodejs.org/api/events.html#events_class_eventemitter) with separate event names per operation type

```javascript
const emitter = collection.watch();
// Register callbacks to react to changes
// Could also be the more verbose `emitter.addListener("insert", event => {`
emitter.on("insert", event => {
  const document = event.fullDocument;
  // Stop listening conditionally based on some value in a document inserted in the collection
  if (document.stop === true) {
    emitter.close();
  }
});
// Register a callback to react to updates
emitter.on("update", event => {
  const document = event.fullDocument;
  // Stop listening conditionally based on some value in a document inserted in the collection
  if (document.stop === true) {
    emitter.close();
  }
});
// Register a callback to react to errors
// This includes errors from establishing the connection as well as errors parsing the event stream, etc.
emitter.on("error", console.error);
// Potentially a way to register a callback to react to any type of event
emitter.on(event => {
  if (event.operationType === "insert") {
    const document = event.fullDocument;
    // Do whatever with the document ...
  }
});
// After some time - close the emitter when we don't care about events anymore
emitter.close();
```

# Alternatives for extending the API later.

If we choose to add an event emitting variant to this API, I see three options

## 0) Provide a `mode` to watch

Users could provide an additional optional option when calling `collection.watch`.

This is probably my favorite, since we would still support an event emitter and it wouldn't feel as "second class" as option 1 (below).

### Using an Async iterator

Probably the default to limit breaking changes and nudge users into using this pattern.

```javascript
for await (const event of collection.watch({ filter, mode: "iterator" })) {
  // Do whatever with event
}
```

### Using an Event Emitter

```javascript
const emitter = collection.watch({ filter, mode: "emitter" });
// Register a listener
emitter.on(event => {
  // Do whatever with event
});
// When you're done, call close
emitter.close();
```

## 1) Adding a method with a different name

The easiest in terms of development time would be to add a future alternative as a separate method.
Ideas for names could be: `collection.watchEmitting()`, `collection.watchWithEmitter()`, etc.


## 2) Break the API

Simply break the API, only returning an event emitter and abandoning the async iterator all together.